## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#f1bbd6fe-de44-4517-8495-a6831b29397d) |
<a name="f1bbd6fe-de44-4517-8495-a6831b29397d"></a>
### 12487-2c86453/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32727
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 04b80bd75c90:/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ol96bna629q5uo",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365579,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16874AE2D461B591, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/16) Running aws-sdk-java tests ... FAILED in 2 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ol96bna629q5uo",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365579,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16874AE2D461B591, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/16) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ol96bna629q5uo",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365579,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16874AE2D461B591, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/16) Running aws-sdk-ruby tests ... FAILED in 3 seconds
{
  "name": "aws-sdk-ruby",
  "function": "listBuckets",
  "args": {},
  "duration": 2142,
  "error": "Server not initialized, please try again.",
  "status": "FAIL"
}
(1/16) Running awscli tests ... FAILED in 8 seconds
{
  "name": "awscli",
  "duration": 6009,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}
(1/16) Running healthcheck tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": null,
  "duration": 72,
  "error": "Get \"http://minio-dev6.minio.io:32727/minio/prometheus/metrics\": EOF",
  "function": "testPrometheusEndpoint",
  "message": "GET request to Prometheus endpoint failed",
  "name": "healthcheck",
  "status": "FAIL"
}
(1/16) Running mc tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": null,
  "duration": 72,
  "error": "Get \"http://minio-dev6.minio.io:32727/minio/prometheus/metrics\": EOF",
  "function": "testPrometheusEndpoint",
  "message": "GET request to Prometheus endpoint failed",
  "name": "healthcheck",
  "status": "FAIL"
}
(1/16) Running minio-dotnet tests ... done in 3 seconds
(2/16) Running minio-go tests ... FAILED in 3 seconds
{
  "args": {
    "bucketName": "minio-go-test-abo6agbejot10rnk",
    "region": "eu-west-1"
  },
  "duration": 3297,
  "error": "Server not initialized, please try again.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testMakeBucketErrorV2",
  "status": "FAIL"
}
(2/16) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket()",
  "args": "[basic check]",
  "duration": 250,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = XMinioServerNotInitialized, message = Server not initialized, please try again., bucketName = minio-java-test-ebm5jg, objectName = null, resource = /minio-java-test-ebm5jg, requestId = 16874AE7A8246171, hostId = 564cc094-1937-4eff-9b3d-15e7b17b9d14)\nrequest={method=PUT, url=http://minio-dev6.minio.io:32727/minio-java-test-ebm5jg, headers=Host: minio-dev6.minio.io:32727\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.2.1\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210610T180909Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210610/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 336\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nRetry-After: 120\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16874AE7A8246171\nX-Xss-Protection: 1; mode=block\nDate: Thu, 10 Jun 2021 18:09:09 GMT\n}\n >>> [io.minio.S3Base.execute(S3Base.java:667), io.minio.MinioClient.makeBucket(MinioClient.java:1207), FunctionalTest.testBucketApi(FunctionalTest.java:442), FunctionalTest.testBucketApiCases(FunctionalTest.java:470), FunctionalTest.makeBucket(FunctionalTest.java:501), FunctionalTest.runBucketTests(FunctionalTest.java:3645), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(2/16) Running minio-js tests ... FAILED in 1 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 4,
  "status": "FAIL",
  "error": "S3Error: Server not initialized, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1336:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(2/16) Running minio-py tests ... FAILED in 10 minutes and 1 seconds
{
  "name": "minio-py:test_make_bucket_default_region",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-2464963e-2d9f-4774-97fa-1044c7689caf",
    "location": "default value ('us-east-1')"
  },
  "message": "HTTPConnectionPool(host='minio-dev6.minio.io', port=32727): Max retries exceeded with url: /minio-py-test-2464963e-2d9f-4774-97fa-1044c7689caf (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 169, in test_make_bucket_default_region\n    _call(log_entry, _CLIENT.make_bucket, bucket_name)\n  File \"/mint/run/core/minio-py/tests.py\", line 109, in _call\n    return func(*args, **kwargs)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 638, in make_bucket\n    headers=headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 574, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev6.minio.io', port=32727): Max retries exceeded with url: /minio-py-test-2464963e-2d9f-4774-97fa-1044c7689caf (Caused by ResponseError('too many 503 error responses',))\n",
  "function": "make_bucket(bucket_name, location, object_lock)",
  "duration": 600459
}
(2/16) Running s3cmd tests ... FAILED in 46 seconds
{
  "name": "s3cmd",
  "duration": "45526",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 3 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 6 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 9 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 12 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 15 sec...\nERROR: S3 error: 503 (XMinioServerNotInitialized): Server not initialized, please try again."
}
(2/16) Running s3select tests ... FAILED in 10 minutes and 0 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-2c984df3-a757-4d69-a2c4-53b7dbff80ac"
  },
  "duration": 600561,
  "message": "HTTPConnectionPool(host='minio-dev6.minio.io', port=32727): Max retries exceeded with url: /s3select-test-2c984df3-a757-4d69-a2c4-53b7dbff80ac (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 95, in test_csv_input_custom_quote_char\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 638, in make_bucket\n    headers=headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 574, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev6.minio.io', port=32727): Max retries exceeded with url: /s3select-test-2c984df3-a757-4d69-a2c4-53b7dbff80ac (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(2/16) Running security tests ... done in 0 seconds
(3/16) Running versioning tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "versioning-test-sn5cu2igh1z9re"
  },
  "duration": 365353,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16874C5F6A978971, host id: ",
  "function": "testCreateVersioningBucket",
  "message": "Versioning CreateBucket Failed",
  "name": "versioning",
  "status": "FAIL"
}

Executed 2 out of 16 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/erasure-server-pool.go
+++ b/cmd/erasure-server-pool.go
@@ -234,6 +234,13 @@ func (z *erasureServerPools) getServerPoolsAvailableSpace(ctx context.Context, b
 	return serverPools
 }
 
+// poolObjInfo represents the state of an object per pool
+type poolObjInfo struct {
+	PoolIndex int
+	ObjInfo   ObjectInfo
+	Err       error
+}
+
 // getPoolIdxExisting returns the (first) found object pool index containing an object.
 // If the object exists, but the latest version is a delete marker, the index with it is still returned.
 // If the object does not exist ObjectNotFound error is returned.
@@ -244,35 +251,49 @@ func (z *erasureServerPools) getPoolIdxExisting(ctx context.Context, bucket, obj
 		return 0, nil
 	}
 
-	errs := make([]error, len(z.serverPools))
-	objInfos := make([]ObjectInfo, len(z.serverPools))
+	poolObjInfos := make([]poolObjInfo, len(z.serverPools))
 
 	var wg sync.WaitGroup
 	for i, pool := range z.serverPools {
 		wg.Add(1)
 		go func(i int, pool *erasureSets) {
 			defer wg.Done()
-			objInfos[i], errs[i] = pool.GetObjectInfo(ctx, bucket, object, ObjectOptions{})
+			// remember the pool index, we may sort the slice original index might be lost.
+			pinfo := poolObjInfo{
+				PoolIndex: i,
+			}
+			pinfo.ObjInfo, pinfo.Err = pool.GetObjectInfo(ctx, bucket, object, ObjectOptions{})
+			poolObjInfos[i] = pinfo
 		}(i, pool)
 	}
 	wg.Wait()
 
-	for i, err := range errs {
-		if err != nil && !isErrObjectNotFound(err) {
-			return -1, err
+	// Sort the objInfos such that we always serve latest
+	// this is a defensive change to handle any duplicate
+	// content that may have been created, we always serve
+	// the latest object.
+	sort.Slice(poolObjInfos, func(i, j int) bool {
+		mtime1 := poolObjInfos[i].ObjInfo.ModTime
+		mtime2 := poolObjInfos[j].ObjInfo.ModTime
+		return mtime1.After(mtime2)
+	})
+
+	for _, pinfo := range poolObjInfos {
+		if pinfo.Err != nil && !isErrObjectNotFound(pinfo.Err) {
+			return -1, pinfo.Err
 		}
-		if isErrObjectNotFound(err) {
+		if isErrObjectNotFound(pinfo.Err) {
 			// No object exists or its a delete marker,
 			// check objInfo to confirm.
-			if objInfos[i].DeleteMarker && objInfos[i].Name != "" {
-				return i, nil
+			if pinfo.ObjInfo.DeleteMarker && pinfo.ObjInfo.Name != "" {
+				return pinfo.PoolIndex, nil
 			}
 
 			// objInfo is not valid, truly the object doesn't
 			// exist proceed to next pool.
 			continue
 		}
-		return i, nil
+		return pinfo.PoolIndex, nil
 	}
 
 	return -1, toObjectErr(errFileNotFound, bucket, object)
@@ -600,6 +621,9 @@ func (z *erasureServerPools) GetObjectNInfo(ctx context.Context, bucket, object 
 	grs := make([]*GetObjectReader, len(z.serverPools))
 
 	lockType = noLock // do not take locks at lower levels
+	checkPrecondFn := opts.CheckPrecondFn
+	opts.CheckPrecondFn = nil
+
 	var wg sync.WaitGroup
 	for i, pool := range z.serverPools {
 		wg.Add(1)
@@ -609,6 +633,26 @@ func (z *erasureServerPools) GetObjectNInfo(ctx context.Context, bucket, object 
 		}(i, pool)
 	}
 	wg.Wait()
+
+	// Sort the objInfos such that we always serve latest
+	// this is a defensive change to handle any duplicate
+	// content that may have been created, we always serve
+	// the latest object.
+	less := func(i, j int) bool {
+		var (
+			mtime1 time.Time
+			mtime2 time.Time
+		)
+		if grs[i] != nil {
+			mtime1 = grs[i].ObjInfo.ModTime
+		}
+		if grs[j] != nil {
+			mtime2 = grs[j].ObjInfo.ModTime
+		}
+		return mtime1.After(mtime2)
+	}
+	sort.Slice(errs, less)
+	sort.Slice(grs, less)
 
 	var found = -1
 	for i, err := range errs {
@@ -627,6 +671,18 @@ func (z *erasureServerPools) GetObjectNInfo(ctx context.Context, bucket, object 
 	}
 
 	if found >= 0 {
+		if checkPrecondFn != nil && checkPrecondFn(grs[found].ObjInfo) {
+			for _, grr := range grs {
+				grr.Close()
+			}
+			return nil, PreConditionFailed{}
+		}
+		for i, grr := range grs {
+			if i == found {
+				continue
+			}
+			grr.Close()
+		}
 		return grs[found], nil
 	}
 
@@ -659,7 +715,6 @@ func (z *erasureServerPools) GetObjectInfo(ctx context.Context, bucket, object s
 
 	errs := make([]error, len(z.serverPools))
 	objInfos := make([]ObjectInfo, len(z.serverPools))
-
 	opts.NoLock = true // avoid taking locks at lower levels for multi-pool setups.
 	var wg sync.WaitGroup
 	for i, pool := range z.serverPools {
@@ -670,6 +725,18 @@ func (z *erasureServerPools) GetObjectInfo(ctx context.Context, bucket, object s
 		}(i, pool)
 	}
 	wg.Wait()
+
+	// Sort the objInfos such that we always serve latest
+	// this is a defensive change to handle any duplicate
+	// content that may have been created, we always serve
+	// the latest object.
+	less := func(i, j int) bool {
+		mtime1 := objInfos[i].ModTime
+		mtime2 := objInfos[j].ModTime
+		return mtime1.After(mtime2)
+	}
+	sort.Slice(errs, less)
+	sort.Slice(objInfos, less)
 
 	var found = -1
 	for i, err := range errs {

--- a/cmd/object-api-utils.go
+++ b/cmd/object-api-utils.go
@@ -792,6 +792,9 @@ func NewGetObjectReader(rs *HTTPRangeSpec, oi ObjectInfo, opts ObjectOptions) (
 
 // Close - calls the cleanup actions in reverse order
 func (g *GetObjectReader) Close() error {
+	if g == nil {
+		return nil
+	}
 	// sync.Once is used here to ensure that Close() is
 	// idempotent.
 	g.once.Do(func() {

